source/common/safe_init/target_impl.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <functional>
+
+#include "envoy/safe_init/target.h"
+
+#include "common/common/logger.h"
+
+namespace Envoy {
+namespace SafeInit {
+
+/**
+ * A TargetHandleImpl functions as a weak reference to a TargetImpl. It is how a ManagerImpl safely
+ * tells a target to `initialize` with no guarantees about the target's lifetime.
+ */
+class TargetHandleImpl : public TargetHandle, Logger::Loggable<Logger::Id::init> {
+private:
+  friend class TargetImpl;
+  TargetHandleImpl(absl::string_view handle_name, absl::string_view name,
+                   std::weak_ptr<std::function<void(WatcherHandlePtr)>> fn);
+
+public:
+  // SafeInit::TargetHandle
+  bool initialize(const Watcher& watcher) const override;
+
+private:
+  // Name of the handle (almost always the name of the ManagerImpl calling the target)
+  std::string handle_name_;
+
+  // Name of the target
+  std::string name_;
+
+  // The target's callback function, only called if the weak pointer can be "locked"
+  std::weak_ptr<std::function<void(WatcherHandlePtr)>> fn_;
+};
+
+/**
+ * A TargetImpl is an entity that can be registered with a Manager for initialization. It can only
+ * be invoked through a TargetHandle. Typical usage is as a "mix-in" class:
+ *
+ *   - Inherit publically from TargetImpl;
+ *   - Implement the `initialize` callback method;
+ *   - When the target is initialized (either immediately in `initialize` or sometime later), call
+ *     `ready` to signal the manager.
+ */
+class TargetImpl : public Target, Logger::Loggable<Logger::Id::init> {
+public:
+  /**
+   * @param name a human-readable target name, for logging / debugging.
+   */
+  TargetImpl(absl::string_view name);
+  ~TargetImpl() override;
+
+  // SafeInit::Target
+  absl::string_view name() const override;
+  TargetHandlePtr createHandle(absl::string_view handle_name) const override;
+
+  /**
+   * Initialize this target. Typically called indirectly via a TargetHandle by the ManagerImpl that
+   * the target was registered with.
+   */
+  virtual void initialize() PURE;
+
+  /**
+   * Signal to the init manager that this target has finished initializing. This should ideally
+   * only be called once, after `initialize` was called. Calling it before initialization begins
+   * or after it has already been called before will have no effect.
```

Comment:
If callers aren't supposed to call this way, can we at least ASSERT that is the case?

Reply:
I should actually probably change the comment, since I think it'll be actually called multiple times on purpose in some places. For example, in https://github.com/envoyproxy/envoy/blob/master/source/common/router/rds_impl.cc#L136:

```c++
void RdsRouteConfigSubscription::runInitializeCallbackIfAny() {
  if (initialize_callback_) {
    initialize_callback_();
    initialize_callback_ = nullptr;
  }
}
```

That's called from a whole bunch of places in the file. The same pattern occurs in other implementations of `Init::Target`, so I figured I'd encapsulate it here.